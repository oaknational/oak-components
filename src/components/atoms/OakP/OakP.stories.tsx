import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakP, OakPProps } from "./OakP";

import { typographyArgTypes } from "@/storybook-helpers/typographyStyleHelpers";

/**
 *
 * OakP is a styled p tag.
 * Use the controls to view different font styles.
 *
 */

const meta: Meta<typeof OakP> = {
  component: OakP,
  tags: ["autodocs"],
  title: "components/atoms/OakP",
  argTypes: {
    ...typographyArgTypes,
  },
  parameters: {
    controls: {
      include: [...Object.keys(typographyArgTypes)],
      sort: "none",
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakP>;

export const DefaultOakFlex: Story = (args: Partial<OakPProps>) => (
  <OakP data-testId="OakP-id" {...args}>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
    tempor incididunt ut labore et dolore magna aliqua. Id semper risus in
    hendrerit gravida rutrum quisque non tellus. Nibh praesent tristique magna
    sit. A arcu cursus vitae congue mauris rhoncus aenean. Turpis egestas
    maecenas pharetra convallis posuere morbi leo. Faucibus pulvinar elementum
    integer enim neque volutpat ac tincidunt. Vitae suscipit tellus mauris a
    diam maecenas.
  </OakP>
);

DefaultOakFlex.args = {
  $color: "black",
  $font: "body-1",
};
