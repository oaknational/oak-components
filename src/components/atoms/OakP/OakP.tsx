import styled from "styled-components";

import { colorStyle, ColorStyleProps } from "@/styles/utils/colorStyle";
import { marginStyle, MarginStyleProps } from "@/styles/utils/spacingStyle";
import {
  typographyStyle,
  TypographyStyleProps,
} from "@/styles/utils/typographyStyle";

export type OakPProps = MarginStyleProps &
  TypographyStyleProps &
  ColorStyleProps;

/**
 * Styled `p` (paragraph) component.
 * ## Usage
 * In general, using a `p` as a descendant of `<Body>` should suffice.
 * However, if you want different styles for a particular paragraph,
 * you can use this component to apply additional styles.
 */
export const OakP = styled.p<OakPProps>`
  ${typographyStyle}
  ${colorStyle}
  ${marginStyle}
`;
