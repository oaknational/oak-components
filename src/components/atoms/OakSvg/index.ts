export { OakSvg } from "./OakSvg";
export type { OakSvgProps, OakSvgNames } from "./OakSvg";
