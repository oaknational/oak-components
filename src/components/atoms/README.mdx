# Atoms

## Rules for components in this folder

- Can import other atoms
- Imports from molecules and organisms are forbidden!
- Generic
- Unstyled
- Avoid embedded logic and state
- No sub-components
  (eg. Box)
