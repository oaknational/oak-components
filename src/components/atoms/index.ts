export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
export * from "./OakIcon";
export * from "./OakGrid";
export * from "./OakGridArea";
export * from "./OakThemeProvider";
export * from "./OakMaxWidth";
export * from "./OakCloudinaryImage";
export * from "./OakKbd";
export * from "./OakGlobalStyle";
export * from "./OakAnchorTarget";
export * from "./OakFieldSet";
