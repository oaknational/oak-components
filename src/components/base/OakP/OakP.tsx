import styled from "styled-components";

import { OakTypographyProps } from "../OakTypography";

import { ColorProps, MarginProps, colorStyle, marginStyle } from "@/styles";
import { typographyStyle } from "@/styles/utils/typographyStyle";

export type OakPProps = MarginProps & OakTypographyProps & ColorProps;
/**
 * Styled `p` (paragraph) component.
 * ## Usage
 * In general, using a `p` as a descendant of `<Body>` should suffice.
 * However, if you want different styles for a particular paragraph,
 * you can use this component to apply additional styles.
 */
const OakP = styled.p<OakPProps>`
  ${typographyStyle}
  ${colorStyle}
  ${marginStyle}

  a {
    color: ${(props) =>
      props.theme &&
      props.theme.uiColors &&
      props.theme.uiColors["text-link-active"]};
  }
`;

export default OakP;
