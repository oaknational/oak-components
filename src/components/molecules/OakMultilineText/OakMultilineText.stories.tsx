import React from "react";
import { StoryObj, Meta } from "@storybook/react";

import { OakMultilineText } from "./OakMultilineText";

const meta: Meta<typeof OakMultilineText> = {
  //  "title" is the title of the story and where to look for component in the storybook
  title: "Components/molecules/OakMultilineText",
  component: OakMultilineText,
  tags: ["autodocs"],
  argTypes: {
    singleLine: { control: "boolean" },
    charLimit: { control: "number" },
    allowCarriageReturn: { control: "boolean" },
    disabled: { control: "boolean" },
    placeholder: { control: "text" },
    invalid: { control: "boolean" },
    invalidText: { control: "text" },
    label: { control: "text" },
  },
  parameters: {
    controls: {
      include: [
        "singleLine",
        "charLimit",
        "allowCarriageReturn",
        "disabled",
        "placeholder",
        "invalid",
        "invalidText",
        "label"
      ],
    },
    backgrounds: {
      default: "light",
    },
  },
};

export default meta;

type Story = StoryObj<typeof OakMultilineText>;

export const Default: Story = {
  render: (args) => <OakMultilineText {...args} />,
  args: {
    // Define your component's default props here
    disabled: false,
    placeholder: "Start typing answer...",
    allowCarriageReturn: true,
    charLimit: 200,
    invalid: false,
    label: "test"
  },
};
