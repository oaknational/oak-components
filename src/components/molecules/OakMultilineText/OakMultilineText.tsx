import React, { useState, forwardRef } from "react";
import styled from "styled-components";

import { OakFlex } from "@/components/atoms/OakFlex";
import { OakTextArea, OakTextAreaProps } from "@/components/atoms/OakTextArea";
import { OakSpan, OakP } from "@/components/atoms";
import { parseColor } from "@/styles/helpers/parseColor";
import { OakCombinedSpacingToken } from "@/styles";
import { ResponsiveValues } from "@/styles/utils/responsiveStyle";

// For example you could restyle the OakFlex component by adding the styles to the css template literal below

export type OakMultilineTextProps = {
  /**
   * Height of component
   */
  $height: ResponsiveValues<OakCombinedSpacingToken>;
  /**
   * Maximum number of characters
   */
  charLimit: number;
  /**
   * Placeholder text
   */
  placeholder?: string;
  disabled: boolean;
  ariaLabel?: string;
  invalid?: boolean;
  invalidText?: string;
  value?: string;
  onChange?: (value: string) => void;
  onFocus?: () => void;
  onBlur?: (value: string) => void;
  onError?: (error: string) => void;
};

type StyledOakTextAreaProps = {
  isError?: boolean;
} & OakTextAreaProps;

const StyledOakTextArea = styled(OakTextArea)<StyledOakTextAreaProps>`
  &:focus-visible {
    border-color: ${(props) =>
      props.isError ? parseColor("red") : parseColor("border-primary")};
    color: ${parseColor("text-primary")};
  }

  &:hover {
    background: ${parseColor("bg-neutral")};
    border-color: ${parseColor("border-neutral")};
  }
`;

const UnstyledComponent = forwardRef(
  (
    {
      charLimit,
      placeholder,
      disabled,
      invalid,
      invalidText,
      ariaLabel,
      value,
      onChange,
      onFocus,
      onBlur,
      onError,
    }: OakMultilineTextProps,
    ref?: React.Ref<HTMLTextAreaElement>,
  ) => {
    const [charCount, setCharCount] = useState(Number);
    const [showCharCount, setShowCharCount] = useState(Boolean);

    const charCountWidth = charLimit > 99 ? "all-spacing-10" : "all-spacing-9";

    const handleFocus = () => {
      onFocus && onFocus();
      setShowCharCount(true);
    };

    const handleBlur = (value: string) => {
      onBlur && onBlur(value);
      setShowCharCount(false);
    };

    const handleChange = (value: string) => {
      onChange && onChange(value);
      const charCount = value.length;
      setCharCount(charCount);
    };

    const handlePaste = (value: string) => {
      if (value.length > charLimit) {
        onError && onError("Character limit exceeded");
      }
    };

    return (
      <OakFlex
        $flexDirection={["row", "column"]}
        $gap={["space-between-xs", null]}
      >
        <StyledOakTextArea
          ref={ref}
          value={value}
          onFocus={() => handleFocus()}
          onChange={(e) => handleChange(e.target.value)}
          onBlur={(e) => handleBlur(e.target.value)}
          maxLength={charLimit}
          placeholder={placeholder}
          disabled={disabled}
          $height={["all-spacing-19", "all-spacing-13", "all-spacing-10"]}
          aria-label={ariaLabel}
          $background="bg-primary"
          $color={"text-subdued"}
          $borderRadius={"border-radius-m"}
          $ba={"border-solid-m"}
          $pa={"inner-padding-s"}
          $width="100%"
          $borderColor={"border-neutral-lighter"}
          onPaste={(e) => handlePaste(e.clipboardData.getData("text"))}
        ></StyledOakTextArea>
        {/* Span is inside OakFlex to stop textarea width changing when charCount changes. */}
        {showCharCount && (
          <OakFlex
            $minWidth={[charCountWidth, null]}
            $justifyContent={[null, "flex-end"]}
            $pa={[null, "inner-padding-ssx"]}
            $position={["relative", null]}
          >
            <OakSpan
              $textAlign={"right"}
              aria-label="character count"
              $font={"body-3"}
              $color={"grey60"}
              $position={["absolute", null]}
              $top={["all-spacing-0", null]}
              $right={["all-spacing-0", null]}
            >
              {charCount}/{charLimit}
            </OakSpan>
          </OakFlex>
        )}
        {invalid && invalidText && (
          <OakP $font={"body-2"} $color={"text-error"}>
            {invalidText}
          </OakP>
        )}
      </OakFlex>
    );
  },
);

/**
 *
 * Add the description of the component here and it will appear on the story for the component
 * The following callbacks are available for tracking focus events:
 *
 * ### Callbacks
 * make sure to add descriptions and types for any callbacks for the component
 *
 * NB. We must export a styled component for it to be inheritable
 */
export const OakMultilineText = UnstyledComponent;
