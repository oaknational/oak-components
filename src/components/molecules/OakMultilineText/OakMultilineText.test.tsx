import React from "react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

import { OakMultilineText } from "./OakMultilineText";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { OakP } from "@/components/atoms/OakP";

describe("OakMultilineText", () => {
  it("renders", () => {
    const { getByPlaceholderText } = renderWithTheme(
      <OakMultilineText
        charLimit={200}
        $height="all-spacing-10"
        placeholder="Start typing answer..."
        disabled={false}
        $width="100%"
        id={"1"}
        name="textarea"
      />,
    );
    expect(getByPlaceholderText("Start typing answer...")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <OakMultilineText
        charLimit={200}
        $height="all-spacing-10"
        disabled={false}
        $width="100%"
        id={"1"}
        name="textarea"
      ></OakMultilineText>,
    );
    expect(container).toMatchSnapshot();
  });

  it("shows char count on focus", async () => {
    const { getByRole, getByLabelText } = renderWithTheme(
      <OakMultilineText
        charLimit={200}
        $height="all-spacing-10"
        disabled={false}
        $width="100%"
        id={"1"}
        name="textarea"
      ></OakMultilineText>,
    );
    const textArea = getByRole("textbox");
    expect(textArea).toBeInTheDocument();
    expect(textArea).not.toHaveFocus();

    expect(() => {
      getByLabelText("character count");
    }).toThrow();

    const user = userEvent.setup();
    await user.click(textArea);
    expect(textArea).toHaveFocus();

    expect(getByLabelText("character count")).toBeInTheDocument();
  });

  it("updates text when value changed", async () => {
    const { getByDisplayValue } = renderWithTheme(
      <OakMultilineText
        charLimit={200}
        $height="all-spacing-10"
        disabled={false}
        value="Hello"
        $width="100%"
        id={"1"}
        name="textarea"
      />,
    );
    const textArea = getByDisplayValue("Hello");
    expect(textArea).toBeInTheDocument();
  });

  it("trims text when pasted text length exceeds character limit", async () => {
    const user = userEvent.setup();
    const onError = jest.fn();

    const { getByRole } = renderWithTheme(
      <OakMultilineText
        charLimit={10}
        $height="all-spacing-10"
        disabled={false}
        onError={onError}
        $width="100%"
        id={"1"}
        name="textarea"
      />,
    );

    const textArea = getByRole("textbox");
    expect(textArea).toBeInTheDocument();
    await user.type(textArea, "Hello");
    await user.click(textArea);
    await user.paste("Testing paste");
    expect(textArea).toHaveValue("HelloTesti");
    expect(textArea).toHaveFocus();

    expect(onError).toHaveBeenCalled();
  });

  it("counts characters correctly when text is edited", async () => {
    const user = userEvent.setup();
    const onTextAreaChange = jest.fn();

    const { getByRole, getByLabelText } = renderWithTheme(
      <OakMultilineText
        charLimit={100}
        $height="all-spacing-10"
        disabled={false}
        onTextAreaChange={onTextAreaChange}
        $width="100%"
        id={"1"}
        name="textarea"
      />,
    );

    const textArea = getByRole("textbox");
    await user.type(textArea, "Hello, world!");
    const charCount = getByLabelText("character count");

    await user.click(textArea);
    expect(textArea).toHaveFocus();
    expect(charCount).toHaveTextContent("13/100");

    await user.type(textArea, "{Backspace}");
    expect(textArea).toHaveValue("Hello, world");
    await user.click(textArea);
    expect(textArea).toHaveFocus();
    expect(charCount).toHaveTextContent("12/100");
    expect(onTextAreaChange).toHaveBeenCalledTimes(14);
  });

  it("shows invalid text if invalid is true and invalid text has been set", () => {
    const { getByRole, getByLabelText } = renderWithTheme(
      <OakMultilineText
        charLimit={100}
        $height="all-spacing-10"
        disabled={false}
        invalidText="Invalid text"
        $width="100%"
        id={"1"}
        name="textarea"
      />,
    );

    const textArea = getByRole("textbox");
    expect(textArea).toBeInTheDocument();

    const invalidText = getByLabelText("invalid text message");
    expect(invalidText).toBeInTheDocument();
    expect(invalidText).toHaveTextContent("Invalid text");
  });

  it("calls onBlur when blurred", async () => {
    const user = userEvent.setup();
    const onBlur = jest.fn();
    const { getByRole, getByText } = renderWithTheme(
      <>
        <OakMultilineText
          charLimit={100}
          $height="all-spacing-10"
          disabled={false}
          onBlur={onBlur}
          $width="100%"
          id={"1"}
          name="textarea"
        />
        <OakP>test</OakP>
      </>,
    );

    const textArea = getByRole("textbox");
    const paragraph = getByText("test");

    await user.click(textArea);
    await user.type(textArea, "text");
    await user.click(paragraph);

    expect(onBlur).toHaveBeenCalled();
  });

  it("calls onFocus when focused", async () => {
    const onFocus = jest.fn();
    const user = userEvent.setup();

    const { getByRole, getByText } = renderWithTheme(
      <>
        <OakMultilineText
          charLimit={100}
          $height="all-spacing-10"
          disabled={false}
          onFocus={onFocus}
          $width="100%"
          id={"1"}
          name="textarea"
        />
        ,<OakP>test</OakP>
      </>,
    );

    const textArea = getByRole("textbox");
    const paragraph = getByText("test");

    await user.click(paragraph);
    expect(textArea).not.toHaveFocus();

    await user.click(textArea);
    expect(textArea).toHaveFocus();

    expect(onFocus).toHaveBeenCalledTimes(1);
  });

  it("displays correct character count when text area value is already set", async () => {
    const user = userEvent.setup();
    const { getByRole, getByLabelText } = renderWithTheme(
      <OakMultilineText
        charLimit={100}
        $height="all-spacing-10"
        disabled={false}
        $width="100%"
        value={"Test"}
        id={"1"}
        name="textarea"
      ></OakMultilineText>,
    );

    const textarea = getByRole("textbox");

    await user.click(textarea);
    const charCount = getByLabelText("character count");
    expect(textarea).toHaveFocus();
    expect(charCount).toBeInTheDocument();
    expect(textarea).toHaveTextContent("Test");
    expect(charCount).toHaveTextContent("4/100");
  });

  it.each([
    [true, "test\ntest 2"],
    [false, "testtest 2"],
  ])(
    "allows carriage return only when true",
    async (allowCarriageReturn, expectedValue) => {
      const user = userEvent.setup();
      const { getByRole } = renderWithTheme(
        <OakMultilineText
          charLimit={200}
          data-testid="test"
          disabled={false}
          placeholder="Start typing answer..."
          allowCarriageReturn={allowCarriageReturn}
        />,
      );
      const textArea = getByRole("textbox");
      expect(textArea).toBeInTheDocument();

      await user.click(textArea);
      await user.type(textArea, "test");

      expect(textArea).toHaveValue("test");

      await user.type(textArea, "{Enter}");
      await user.type(textArea, "test 2");
      expect(textArea).toHaveValue(expectedValue);
    },
  );

  it("trims leading and trailing whitespace on blur when allowLeadingTrailingSpaces is false", async () => {
    const user = userEvent.setup();
    const onError = jest.fn();
    const onTextAreaChange = jest.fn();
    const { getByRole } = renderWithTheme(
      <OakMultilineText
        charLimit={200}
        data-testid="test"
        disabled={false}
        onError={onError}
        onTextAreaChange={onTextAreaChange}
        placeholder="Start typing answer..."
        allowLeadingTrailingSpaces={false}
      />,
    );
    const textArea = getByRole("textbox");
    expect(textArea).toBeInTheDocument();

    await user.click(textArea);
    await user.type(textArea, "   test   ");

    expect(onTextAreaChange).toHaveBeenLastCalledWith("   test   ");
    await user.click(document.body); // blur

    expect(onTextAreaChange).toHaveBeenLastCalledWith("test");
    expect(onError).toHaveBeenCalledWith("Forbidden characters in input");
  });

  it("removes carriage returns on pasted text", async () => {
    const user = userEvent.setup();
    const onError = jest.fn();
    const onTextAreaChange = jest.fn();
    const { getByRole } = renderWithTheme(
      <OakMultilineText
        charLimit={200}
        data-testid="test"
        disabled={false}
        onError={onError}
        onTextAreaChange={onTextAreaChange}
        placeholder="Start typing answer..."
        allowLeadingTrailingSpaces={true}
      />,
    );
    const textArea = getByRole("textbox");
    expect(textArea).toBeInTheDocument();

    await user.click(textArea);
    await user.paste("test\ntest 2");

    expect(onTextAreaChange).toHaveBeenLastCalledWith("test test 2");
    expect(onError).toHaveBeenCalledWith("Forbidden characters in input");
  });
});
