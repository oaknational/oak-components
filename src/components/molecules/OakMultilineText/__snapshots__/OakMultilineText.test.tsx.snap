// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`OakMultilineText matches snapshot 1`] = `
.c0 {
  width: 100%;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0 c1 "
  disabled={false}
  placeholder="Start typing answer..."
>
  <textarea
    disabled={false}
    id="OakMultilineText"
    maxLength={200}
    onKeyDown={[Function]}
    placeholder="Start typing answer..."
    rows={4}
    style={
      {
        "overflow": "visible",
        "resize": "none",
      }
    }
  />
</div>
`;
